soroban-sdk/src/auth.rs

```diff
@@ -13,3 +15,21 @@ impl Context {
         Address::from_contract_id(&self.contract)
     }
 }
+
+/// Auth interface that a contract implements to support being used for custom
+/// auth.
+///
+/// Once a contract implements the interface, the address of the contract can be
+/// used for addresses that have [`Address::require_auth`] called.
+pub trait Interface<E>
+where
+    Status: for<'a> TryFrom<&'a E>,
+{
+    /// Check that the signatures and auth contexts are valid.
+    fn __check_auth(
+        env: Env,
+        signature_payload: BytesN<32>,
+        signatures: Vec<RawVal>,
```

Comment:
One interesting caveat is that without the interface it's possible to do type auto-conversion via SDK and have something like `Vec<BytesN<64>>` or whatever appropriate UDT. Probably not a big deal though.

Reply:
That's a good point. I changed the interface to have an associated type used for the signature type.